### PR TITLE
Move routing table exchange to another thread

### LINKED ISSUE #4859: Move old exchange routing table to new thread
After we push https://github.com/near/nearcore/pull/4112, routing table information will be duplicated. 

We should move computing old routing table to new thread, this will offload computing routing table from PeerManager thread.

- [x] Implement offloading, to another thread, sending routing table to another peer (https://github.com/near/nearcore/pull/4886)
- [x] Move RoutingTable (  pub raw_graph: Graph,) to another thread. (https://github.com/near/nearcore/pull/5089)
- [x] Remove copy of edges_info routing.rs, and keep only one in RoutingTableActor.  (https://github.com/near/nearcore/pull/5089)

### PR DESCRIPTION
**Pull request by pmnoxx**
In this PR we will creating copy of routing table to another thread.

This is related to https://github.com/near/nearcore/issues/4859

Summary:
- `PeerManager` no longer has copy of `edges_info`, instead it only stores list of edges to peers that it is/was connected to `local_edges_info`
- Removed usage of edges in `PeerManager` that were stored in ` edges_info`
- Computation of `peer_forwarding` is done in `RoutingTableActor`, `PeerManager` gets it from another thread every second / tests now use `RoutingTableActor` instead of `RoutingTable` struct
- moved `pub raw_graph: Graph,`  and a few other data structures  to `RoutingTableActor`
- Removed `adv_get_routing_table_new` from jsonrpc server. Now, routing table is only stored in one place, so this method is no longer needed
- Added code/struct comments

Closes #4859

### CODE REVIEW

**Review by bowenwang1996 — COMMENTED** [review]
@mm-near please review

**Comment by pmnoxx** [comment]
@mm-near If you think it will make reviewing easier, I can extract the refactoring mentioned in summary to a separate change.

**Comment by pmnoxx** [comment]
@mm-near @bowenwang1996 I added summary to the first post in the PR, I hope it helps.

File: chain/network/src/peer_manager.rs
Comment (matklad) [inline chain/network/src/peer_manager.rs:312]: thought: curiously, this essentially duplicates `(peer0, peer1)` twice, as they are stored both as keys and as values. With some rust trickery, it should be possible to store just a `HashSet<Edge>`, which is then query-able using a `&(Peer, Peer)` pair (you'll need to implment `Borrow<(PeerId, PeerId)> for Edge`). Probably doesn't makes sense here, but a good trick to keep in mind.

File: chain/network/src/peer_manager.rs
Comment (mm-near) [inline chain/network/src/peer_manager.rs:197]: for future: we should try to make sure that variable names (and meanings are matching)

here I'd rename it to self.adv_can_broadcast_edges

File: chain/network/src/peer_manager.rs
Comment (mm-near) [inline chain/network/src/peer_manager.rs:313]: looking at the code, I don't think that this part belongs to this function.

pending_update_nonce_request is the map that holds the last "nonce" that we sent to one of our direct neighbours about the edge that we have with them (or to be exact, about the case where this edge is about to be removed).

And this piece of code is handling the "successful" response/confirmation.

So maybe we should put these lines (314-318) back to 1899
and in there (1899) just call the add_verified_edges_to_routing_table with mode "default" ?

File: chain/network/src/routing.rs
Comment (pmnoxx) [reply chain/network/src/routing.rs]: This 3 lines will disappear after rebase.  See https://github.com/near/nearcore/pull/5102

File: chain/network/src/routing.rs
Comment (pmnoxx) [reply chain/network/src/routing.rs:412]: That's a good point, we were using `me` in a few places.

File: chain/network/src/routing.rs
Comment (pmnoxx) [reply chain/network/src/routing.rs:435]: Thanks, done.

File: chain/network/src/peer_manager.rs
Comment (pmnoxx) [reply chain/network/src/peer_manager.rs:127]: I updated code to be "my_peer_id" in all places for consistency.

File: chain/network/src/peer_manager.rs
Comment (pmnoxx) [reply chain/network/src/peer_manager.rs:313]: Ok, done

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs:59]: Added comment, and renamed to 'needs_routing_table_recalculation'

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs]: Ok, I added comment, and renamed `add_edge` to `add_verified_edge`

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs:63]: yes

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs:213]: Edge to remove, I refactor the code.

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs:599]: Ok.

File: chain/network/src/peer_manager.rs
Comment (pmnoxx) [reply chain/network/src/peer_manager.rs:196]: Ok

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs:327]: Changed comment to "Get edges stored in DB under `ColPeerComponent` column at `peer_id` key."

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs:196]: You are right, if there is a crash in the middle we will lose those edges. That's true for all active edges in memory, they are not stored anywhere.

I don't think this is an issue, but rather than odd behavior.

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs]: Yes, this is a bug. Fixed it https://buildkite.com/nearprotocol/nearcore/builds/5943

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs:286]: I added comments about how we store data in DB.

File: chain/network/src/peer_manager.rs
Comment (pmnoxx) [reply chain/network/src/peer_manager.rs:300]: It's good to know that's it's possible. We could save even more memory by storing `Arc<PeerId>` and have an extra data structure, which removed duplicates.

For example `HashSet<Arc<PeerId>>` inside `RoutingTableActor`. I'm not sure if it's worth the effort.

File: chain/network/src/peer_manager.rs
Comment (pmnoxx) [reply chain/network/src/peer_manager.rs:262]: That's a good point. Though the signature for the method would look like prune: Option<Pair<bool, Duration>>. I think keeping those as separate values is more readable, but yes, I see your point. I'll keep it as it is for now.

**Comment by pmnoxx** [comment]
@bowenwang1996 @mm-near Is there anything stopping us from merging this PR?

**Review by mm-near — CHANGES_REQUESTED** [review]

**Comment by mm-near** [comment]
@pmnoxx  - sorry, I had pending comments but I didn't click submit (darn github UI :-)

From my perspective - I need to do the "final full check" (that would include looking at the tests etc) - and then we'll be good to go.

**Comment by mm-near** [comment]
and hopefully you can submit the other PR #5140 - which should make this diff smaller.

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs:41]: Thanks, fixed.

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs:137]: Yes, I updated the comment.

**Comment by bowenwang1996** [comment]
@pmnoxx looks like CI does not pass. Also you need approval from @mm-near

**Comment by pmnoxx** [comment]
> @pmnoxx looks like CI does not pass. Also you need approval from @mm-near

Thanks, fixed.

File: chain/network/src/routing_table_actor.rs
Comment (pmnoxx) [reply chain/network/src/routing_table_actor.rs]: Fixed

File: pytest/tests/sanity/routing_table_sync.py
Comment (bowenwang1996) [inline pytest/tests/sanity/routing_table_sync.py:181]: Why do we remove this function? It seems that we removed part of the test. Is that intentional?

**Review by bowenwang1996 — COMMENTED** [review]
@mm-near please review this PR

File: pytest/tests/sanity/routing_table_sync.py
Comment (pmnoxx) [reply pytest/tests/sanity/routing_table_sync.py:181]: We used to have copy of routing table in two places (`PeerManagerActor`, `RoutingTableActor`). After we merge this PR we will have only one place with routing table. So, the second method is no longer needed.

**Comment by pmnoxx** [comment]
@mm-near Can you take a look? This PR is much smaller after splitting it 6 times.

File: chain/network/src/peer_manager/peer_manager_actor.rs
Comment (mm-near) [inline chain/network/src/peer_manager/peer_manager_actor.rs:348]: comment

File: chain/network/src/peer_manager/peer_manager_actor.rs
Comment (mm-near) [inline chain/network/src/peer_manager/peer_manager_actor.rs:413]: ... and broadcast them to the peers

File: chain/network/src/peer_manager/peer_manager_actor.rs
Comment (mm-near) [inline chain/network/src/peer_manager/peer_manager_actor.rs:380]: can you comment what this whole for loop does?

It seems to be some sort of reconciliation mechanism - we're looking at the new edges that are added to the system, focusing only on the ones that are pointing to us - and we try to match this to our internal state.

File: chain/network/src/peer_manager/peer_manager_actor.rs
Comment (mm-near) [inline chain/network/src/peer_manager/peer_manager_actor.rs:1531]: comment: update routing table and prune edges that are no longer available.

**Review by mm-near — APPROVED** [review]

**Comment by pmnoxx** [comment]
@chefsale @frol @khorolets Can you take a look?

File: chain/network/src/peer_manager/peer_manager_actor.rs
Comment (pmnoxx) [reply chain/network/src/peer_manager/peer_manager_actor.rs:390]: Thanks, I added a comment:

```
            // Check whenever there is an edge indicating whenever there is a peer we should be
            // connected to but we aren't. And try to resolve the inconsistency.
            //
            // Also check whenever there is an edge indicating that we should be disconnected
            // from a peer, but we are connected. And try to resolve the inconsistency.
```

File: chain/network/src/peer_manager/peer_manager_actor.rs
Comment (pmnoxx) [reply chain/network/src/peer_manager/peer_manager_actor.rs:1545]: Thanks fixed. We are pruning edges that are no longer reachable. Those may be existing edges, but to which we can't find a valid path.

Comment (pmnoxx) [reply chain/network/src/peer_manager/peer_manager_actor.rs:1546]:
```
        // Update routing table and prune edges that are no longer reachable.
        self.update_routing_table_trigger(ctx);
```

**Comment by pmnoxx** [comment]
@chefsale @frol @khorolets Can you take a look?

**Review by chefsale — APPROVED** [review]

**Pull request merged**
